src/main/java/jewellery/inventory/service/ProductService.java

```diff
@@ -0,0 +1,230 @@
+            throwExceptionWhenResourceInUserNotExist(user, resourceInUser, resource, resourcesInUsers);
+
+            double quantity = resourceQuantityRequestDto.getQuantity();
+            setResourcesToProduct(resourceInUser, quantity, resourceInProduct, resource, resourcesInProducts);
+
+            removeResourceFromUser(user, resourceInUser);
+
+        });
+
+        return resourcesInProducts;
+    }
+
+    private void throwExceptionWhenResourceInUserNotExist(User user, ResourceInUser resourceInUser, Resource resource, List<ResourceInUser> resourcesInUsers) {
+
+        if (resourceInUser == null || !resourcesInUsers.contains(resourceInUser)) {
+            throw new ResourceInUserNotFoundException(resource.getId(), user.getId());
+        }
+    }
+
+    private void setResourcesToProduct(ResourceInUser resourceInUser, double quantity, ResourceInProduct resourceInProduct, Resource resource, List<ResourceInProduct> resourcesInProducts) {
+        if (resourceInUser.getQuantity() < quantity) {
+            throw new InsufficientResourceQuantityException(quantity, resourceInUser.getQuantity());
+        } else {
+            resourceInProduct.setResource(resource);
+            resourceInProduct.setQuantity(quantity);
+            resourcesInProducts.add(resourceInProduct);
+            resourceInUser.setQuantity(resourceInUser.getQuantity() - quantity);
+        }
+    }
+
+    private void removeResourceFromUser(User user, ResourceInUser resourceInUser) {
+        if (resourceInUser.getQuantity() == 0) {
+            resourceInUserService.removeResourceFromUser(user.getId(), resourceInUser.getResource().getId());
+        }
+    }
+
+    private List<Product> getProductsInProduct(List<UUID> productsIdInRequest, Product parentProduct) {
+        List<Product> products = new ArrayList<>();
+        if (productsIdInRequest != null) {
+            productsIdInRequest.forEach(productId -> {
+                Product product = productRepository.findById(productId)
+                        .orElseThrow(() -> new ProductNotFoundException(productId));
+                product.setContentOf(parentProduct);
+                products.add(product);
+            });
+        }
+
+        return products;
+    }
+
+    private ResourceInUser moveResourcesToUserWhenDisassembleProduct(User owner, ResourceInProduct resourceInProduct, Resource resource) {
+        ResourceInUser resourceInUser = resourceInUserRepository.findByResourceId(resource.getId());
+        if (resourceInUser == null) {
+            resourceInUser = new ResourceInUser();
+            resourceInUser.setResource(resource);
+            resourceInUser.setQuantity(resourceInProduct.getQuantity());
+            resourceInUser.setOwner(owner);
+        } else {
+            resourceInUser.setQuantity(resourceInUser.getQuantity() + resourceInProduct.getQuantity());
+        }
+        return resourceInUser;
+    }
```

Comment:
If we utilise ResourceInUserService::addResourceToUser, then this method would be obsolete.